From 6f4fd28fade3d1e2903e3cb47a831fca2d13e9a4 Mon Sep 17 00:00:00 2001
From: Greg Moffatt <greg.moffatt@windriver.com>
Date: Tue, 24 Aug 2010 17:08:39 -0400
Subject: [PATCH] sctp: fix oops with unconnected sockets

In a few boundary cases a socket could be used where it isn't in a
connected or listening state for SCTP traffic.  When this occurs
the attached network is NULL which in-turn causes the __xfrm_lookup()
function to panic by dereferencing the NULL net.

To address this a check is added in the SCTP get_dst() function where
if a net attached to a socket is NULL, the function simply returns
NULL indicating no destination for the intended SCTP association as
the socket itself has no connection.

Signed-off-by: Greg Moffatt <greg.moffatt@windriver.com>
---
 net/sctp/ipv6.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/net/sctp/ipv6.c b/net/sctp/ipv6.c
index e7e6e2d..48b2b3e 100644
--- a/net/sctp/ipv6.c
+++ b/net/sctp/ipv6.c
@@ -278,6 +278,8 @@ static struct dst_entry *sctp_v6_get_dst(struct sctp_association *asoc,
 			if (sk)
 #endif
 			net = sock_net(sk);
+			if (!net)
+				goto failure;
 			fl.proto = IPPROTO_SCTP;
 			if (sk) fl.oif = sk->sk_bound_dev_if;
 			fl.fl_ip_dport = daddr->v6.sin6_port;
-- 
1.6.5.2

